trainer.py

```diff
@@ -0,0 +1,133 @@
+from battler import Battler
+from model.model_actor import ModelActor
+from player_actor import RandomActor
+from utils import simulate_games
+import torch
+import torch.optim as optim
+
+def train(model_actor, random_actor, gamma, num_episodes, optimizer):
+    score = {'BOT_1': 0, 'BOT_2': 0}
+    
+    ############## Implementing Policy Gradients (REINFORCE ALGORITHM) below ##############
+
+    # Training Loop
+    for episode in range(num_episodes):
+        battler = Battler(model_actor, random_actor)
+        battler.make_moves() # Initial game setup hack so p1move and p2move are not both False
+
+        training_samples = [] # store relevant metadata from a single episode / battle
+
+        ############## PLAY AN ENTIRE EPISODE / BATTLE AND RECORD EACH INTERACTION ##############
+        while battler.current_state != 'end':
+            prob_action = None # probability of action chosen by model in current game state
+            reward = None # reward received after executing actions in current game state
+
+            original_total_hp_actor_1 = model_actor.team.calculate_total_HP()
+            original_total_hp_actor_2 = battler.actor2.team.calculate_total_HP()
+            
+            if battler.p1move:
+                knowledge = {"field": None, "opponent": battler.actor2.team, "error": battler.error}
+                action = model_actor.pick_move(knowledge)
+
+                action_type, action_num = action.split(" ")
+                action_index = None
+                if (action_type == "move"):
+                    action_index = int(action_num) - 1
+                else:
+                    action_index = int(action_num) + 2
+
+                prob_action = model_actor.prev_probs[0][action_index]
+                # Earlier, I noticed that model's prev_probs is NAN during some runs (model updating weights in strange ways perhaps?)
+                # Anyway, this issue appears to be mitigated for now, but leaving this check here for the future!
+                if (torch.isnan(model_actor.prev_probs).any()):
+                    print("Model's prev_probs contains NAN values")
+
+                move = f">p1 {action}"
+                # print(move)
+                battler.send_command(move)
+                battler.p1move = False
+
+            if battler.p2move:
+                knowledge = {"field": None, "opponent": model_actor.team, "error": battler.error}
+                move = f">p2 {battler.actor2.pick_move(knowledge)}"
+                # print(move)
+                battler.send_command(move)
+                battler.p2move = False
+            
+            battler.receive_output()
+
+            # TODO: Improve reward heuristic
+            # Calculate reward based on damage exchanged (considering full teams in case a switch move occurred)
+            damage_dealt = battler.actor2.team.calculate_total_HP() - original_total_hp_actor_2
```

Comment:
It just occurred to me- is this backwards?

ex. Both us and the opponent start with 100 total hp. We do 5 points of damage to them, we don't lose any health (good!).

damage_dealt = 95 - 100 = -5
damage_recieved = 0
reward = -5 (or now .1)

Similarly if we receive more damage than we deal then the reward is a positive number, which again seems backwards.